### PR TITLE
ASoC: SOF: protect access to trace offset with locks

### LINKED ISSUE #655: Upstream: trace: clarify use of host_offset
Feedback from Takashi Iwai

````
> +
> +	/* return bytes available for copy */
> +	if (sdev->host_offset < pos)
> +		return buffer_size - pos;
> +
> +	return sdev->host_offset - pos;

Are these sdev->host_offset accesses race-free?

If I understand correctly, snd_sof_trace_update_pos() can be called at
any time, and the offset might become inconsistent during these
multiple references.
````

### PR DESCRIPTION
**Pull request by kv2019i**
Fixes #655 

The debugfs trace offset is accessed both by debugfs trace
client as well as the IPC irq handler. Add locking for all
accesses to the trace offset. This ensures free space checks
done by debugfs client always have a consistent view to trace
buffer state.

The field name was renamed from 'trace_offset' to align
with the lock name and actual usage in debugfs trace code.

Signed-off-by: Kai Vehmanen <kai.vehmanen@linux.intel.com>

### CODE REVIEW

**Comment by ranj063** [comment]
@kv2019i seems like there's a couple of checkpatch issues to resolve

**Comment by plbossart** [comment]
> @kv2019i seems like there's a couple of checkpatch issues to resolve

yes, it's worth fixing these two:

````
CHECK: spinlock_t definition without comment
#45: FILE: sound/soc/sof/sof-priv.h:408:
+	spinlock_t trace_lock;
CHECK: Alignment should match open parenthesis
#137: FILE: sound/soc/sof/trace.c:264:
+	if (sdev->dtrace_is_enabled &&
+		sdev->trace_offset != posn->host_offset) {
````

File: sound/soc/sof/trace.c [hunk 1]
```diff
@@ -17,20 +17,29 @@ static size_t sof_wait_trace_avail(struct snd_sof_dev *sdev,
 				   loff_t pos, size_t buffer_size)
 {
 	wait_queue_entry_t wait;
+	unsigned long flags;
+	size_t ret;
+
+	spin_lock_irqsave(&sdev->trace_lock, flags);
```
Comment (lyakh) [inline sound/soc/sof/trace.c]: This function cannot (or at least must not) be called from an atomic context, so, there's no need to save IRQ flags, IRQs aren't disabled here. Therefore spin_(un)lock_irq() would suffice.

**Comment by kv2019i** [comment]
>> @kv2019i seems like there's a couple of checkpatch issues to resolve
> yes, it's worth fixing these two:

Yikes! To my defense :), I did run checkpatch locally, but I did not use --strict. Note taken, will fix the patch and adapt my local scripts to this.

**Comment by lyakh** [comment]
Sorry, I don't see what this is solving and how. You're adding a lock inside sof_wait_trace_avail() to get a "consistent view" on the buffer. But it seems to me, that you're only locking the read of .host_offset (or .trace_offset after your renaming), so, I don't understand - you're trying to keep your view on it consistent with what? What should be done here is protecting access to the buffer together with reading of the counter. But the buffer is read out outside of that function when the lock has been long dropped. I think, to really protect the buffer you have to take the lock, check the counter *and* copy out the desired data to a temporary buffer. Then you can release the lock and copy the data from that temporary buffer to the user. Without that after you have released your lock in that function, a trace event can arrive, the buffer contents is changed and you get corrupted data out.

In fact I don't think you can really protect yourself against getting corrupted data, unless you drop messages when the buffer is full. As far as I understand we keep getting trace data from the DSP wrapping around as the end of the buffer is reached. If the host didn't read the buffer quickly enough, it will get out of sync anyway, even if we protect our side perfectly. Our current reading position will be overwritten by the middle of a random new message and we'll get out of sync. Unless the IPC message size is fixed and is a power of 2?

**Comment by kv2019i** [comment]
@lyakh wrote:
> Sorry, I don't see what this is solving and how. You're adding a lock inside sof_wait_trace_avail() to get > a "consistent view" on the buffer. But it seems to me, that you're only locking the read of .host_offset 
> (or .trace_offset after your renaming), so, I don't understand - you're trying to keep your view on it 
> consistent with what? 

This is only protecting modifications and reads of the trace_offset. Without the lock, a softirq may interrupt user process in middle of sof_wait_trace_avail(). Specifically we may hit:

```
»       if (sdev->trace_offset < pos)
»       »       return buffer_size - pos;
<<PREEMPT to irq handler - trace_offsets wraps over>>
»       if (sdev->trace_offset > pos)
»       »       return sdev->trace_offset - pos;
```

This is a bit theoretical as sdev->trace_offset is likely read only once from memory, but with spinlock we are making it clear. And the wrap-around will happen frequently in the 64KB ring buffer.

> What should be done here is protecting access to the buffer together with reading of the counter. 
> But the buffer is read out outside of that function when the lock has been long dropped. I think, to 
> really protect the buffer you have to take the lock, check the counter _and_ copy out the desired data > to a temporary buffer. Then you can release the lock and copy the data from that temporary buffer to 

Ack, this does not protect the buffer -- this would require a bigger rework of the current design that is based on the shared 64K ringbuffer. If user-space falls behind so that DSP write pointer reaches user-space read ptr, it will just silently overwrite it as we have no other buffer for the traces at the moment. Sof-logger will try to resync by reading a word at a time if is unable to parse so it can recover, but you will get inconsistent traces out at start.

But in the typical debug usage where you are running "sof-logger -t", you will not hit this problem with _buffer_ consistency. It is however possible to hit the trace_offset update race, even if user-space is reading data all the time, so this patch does address that aspect.

I've been looking at fs/seq_file.c. That could be a longer term solution to implement the debugfs trace. That will add a separate buffer and use common kernel code to manage it (this is what kernel/trace uses).

**Comment by lyakh** [comment]
> @lyakh wrote:
> 
> > Sorry, I don't see what this is solving and how. You're adding a lock inside sof_wait_trace_avail() to get > a "consistent view" on the buffer. But it seems to me, that you're only locking the read of .host_offset
> > (or .trace_offset after your renaming), so, I don't understand - you're trying to keep your view on it
> > consistent with what?
> 
> This is only protecting modifications and reads of the trace_offset. Without the lock, a softirq may interrupt user process in middle of sof_wait_trace_avail(). Specifically we may hit:
> 
> ```
> »       if (sdev->trace_offset < pos)
> »       »       return buffer_size - pos;
> <<PREEMPT to irq handler - trace_offsets wraps over>>
> »       if (sdev->trace_offset > pos)
> »       »       return sdev->trace_offset - pos;
> ```

@kv2019i Ok, so, what you're trying to protect against here is, that the first check `sdev->trace_offset < pos` returns false because at that time `sdev->trace_offset >= pos` but before you get to the next check for `sdev->trace_offset > pos` the buffer wraps and now already `sdev->trace_offset < pos` is true again, but instead of taking the data you enter waiting. You will be woken up by the next IPC anyway, but kind of it isn't good. Wouldn't just using a local variable fix that? Like
```
size_t offset = sdev->trace_offset;
if (offset < pos)
	return buffer_size - pos;
if (offset > pos)
	return offset - pos;
```
And then do one more read into a local variable after the sleep?

**Comment by kv2019i** [comment]
@lyakh wrote:
> @kv2019i Ok, so, what you're trying to protect against here is, that the first check `sdev->trace_offset > < pos` returns false because at that time `sdev->trace_offset >= pos` but before you get to the next 
> check for `sdev->trace_offset > pos` the buffer wraps and now already `sdev->trace_offset < pos` is 
[...]
> And then do one more read into a local variable after the sleep?

And I was jut confirming by looking at gcc output that even with current code, sdev->host_offset is only read once before sleep in our current platforms (gcc and x86_64),  so in this sense local variable would not change practical output.

```
size_t offset = sdev->trace_offset;
if (offset < pos)
	return buffer_size - pos;
if (offset > pos)
	return offset - pos;
```
What bugs me in this approach is that we are relying on gcc/compiler behaviour here. It is not strictly limited to reading sdev->trace_offset only once, even if we have the local variable. Agreed I can't think of a case where this would happen, so maybe this is a too theoretical concern.

Hmm, this popped in Takashi's review in any case, so maybe a middleground would be to add the local variable to state the intent of the code better. Spinlocks might set the expectations too high -- the buffer is still best effort with current design (and this maybe perfectly ok for debug). I'll update the patch.

**Comment by lyakh** [comment]
> And I was jut confirming by looking at gcc output that even with current code, sdev->host_offset is only read once before sleep in our current platforms (gcc and x86_64), so in this sense local variable would not change practical output.

Well, relying on that isn't good, as you said - that would be relying on some specific "random" compiler behaviour.

> What bugs me in this approach is that we are relying on gcc/compiler behaviour here. It is not strictly limited to reading sdev->trace_offset only once, even if we have the local variable. Agreed I can't think of a case where this would happen, so maybe this is a too theoretical concern.

I actually don't think it would be valid for the compiler to re-read the object multiple times instead of using a local variable in this case for exactly this reason - the object can change outside of this code. And I don't even think you have to declare the pointer as volatile for that.

**Comment by kv2019i** [comment]
@lyakh wrote:
> I actually don't think it would be valid for the compiler to re-read the object multiple times instead 
> of using a local variable in this case for exactly this reason - the object can change outside of this code. > And I don't even think you have to declare the pointer as volatile for that.

Why not? At least I couldn't find any solid reference for this. Granted a deference is more expensive, so compiler has no reason to do this unless there's a volatile keyword for the object. Without volatile, compiler doesn't care about outside changes, so to me it would seem it is ok to do this. There's bunch of macros like READ_ONCE() and ACCESS_ONCE() that seem to address this in linux/compiler.h (and see https://lwn.net/Articles/624126/ ).

Hmm, atomics is one, but that's probably overkill here as well (the offset is 32bit after all). So let's do so I'll add a local variable and READ_ONCE(). How about that?

File: sound/soc/sof/trace.c (hunk 1 above)
Comment (kv2019i) [reply sound/soc/sof/trace.c]: @lyakh I will now remove the spinlock, but as an educational comment: I actually had that in first version, but then I looked at existing sof ipc.c code and that seems to be using irqsave() variants, so I aligned with that. I was wondering about the reason for this -- maybe to make these functions callable both from atomic and non-atomic context (these are in SOF core code after all). Or is this just wrong in ipc.c code as well?

**Comment by lyakh** [comment]
> @lyakh wrote:
> 
> > I actually don't think it would be valid for the compiler to re-read the object multiple times instead
> > of using a local variable in this case for exactly this reason - the object can change outside of this code. > And I don't even think you have to declare the pointer as volatile for that.
> 
> Why not? At least I couldn't find any solid reference for this. Granted a deference is more expensive, so compiler has no reason to do this unless there's a volatile keyword for the object. Without volatile, compiler doesn't care about outside changes, so to me it would seem it is ok to do this. There's bunch of macros like READ_ONCE() and ACCESS_ONCE() that seem to address this in linux/compiler.h (and see https://lwn.net/Articles/624126/ ).
> 
> Hmm, atomics is one, but that's probably overkill here as well (the offset is 32bit after all). So let's do so I'll add a local variable and READ_ONCE(). How about that?

@kv2019i Wow, very cool! Indeed, I wasn't 100% certain either, but it seem logical to me, that dereferencing pointers multiple times in contradiction to the source code would be disallowed. I stand corrected! Very nice articles, thanks, lett's use READ_ONCE()!

File: sound/soc/sof/trace.c [hunk 2]
```diff
@@ -17,18 +17,19 @@ static size_t sof_wait_trace_avail(struct snd_sof_dev *sdev,
 				   loff_t pos, size_t buffer_size)
 {
 	wait_queue_entry_t wait;
+	u32 host_offset = READ_ONCE(sdev->host_offset);
```
Comment (lyakh) [inline sound/soc/sof/trace.c]: /me looking for a "llayk" button :-D but... the "u32" type makes be cringe just a bit... I understand that it's selected to match sdev->host_offset but that variable is then compared and added to pos, which is loff_t... And I think loff_t is a more appropriate type for this and ot actually would be good to change that in struct snd_sof_dev too...

Comment (kv2019i) [reply sound/soc/sof/trace.c]: > /me looking for a "llayk" button :-D but... the "u32" type makes be cringe just a bit... I understand that it's selected to match sdev->host_offset but that variable is then compared and added to pos, which is loff_t... And I think loff_t is a more appropriate type for this and ot actually would be good to change that in struct snd_sof_dev too...

Hmm, loff_t is probably better. I don't want to go and touch snd_sof_dev instance at least yet. If we make the counter 64bit, then we open a new can of worms as we cannot assume 64bit read/writes are atomic across platforms (although they are on x86_64) and need to move to atomics. In generic C, one can't assume that for 32bit read/writes either, but I believe in Linux kernel this is ok and common practise.

**Comment by plbossart** [comment]
@lyakh let me know when you are ok with this.

**Comment by lyakh** [comment]
> @lyakh let me know when you are ok with this.

@plbossart now

**Comment by plbossart** [comment]
@xiulipan again failed CI. the code was pushed 2 hours ago and there's no build results.

**Comment by plbossart** [comment]
@wenqingfu I have other things to do today so will merge after checking that the code compiled locally. we really need to fix for those CI issues.

**Comment by xiulipan** [comment]
@plbossart @wenqingfu 
Agreed. We may need some maintainer in different time zone to check the status of the CI.

**Pull request merged**